configure.ac

```diff
@@ -95,6 +95,22 @@ else
             [Defined if we're building with OpenSSL or LibreSSL])
 fi
 
+dnl Enable Android only features.
+AC_ARG_ENABLE(android,
+     AS_HELP_STRING(--enable-android, [setup build for Android exeutable/shared library]))
+
+if test "x$enable_android" = "xyes"; then
+  asciidoc=false
```

Comment:
I'm okay with overriding these defaults, but we need to test to make sure that the user can override them if they want. So I think we need to make sure (and document) that this comes before any of the other options that it adjusts.  Does it?

Reply:
If someone wants to customize all those things, they can omit `--enable-android`, then specify them one at a time.

After talking with Orbot, Briar, Thali, Tor Browser, and iOS OnionBrowser developers on the build config topic, and looking at the relevant bits in the code of those projects, I can see no real use case for tweaking the `--enable-android` flag.  This isn't like UNIX, where there are different packages and libraries installed.  Tor on Android and iOS it built by the native wrappers using `--enable-android`. 

There is never going to be systemd or system-torrc on Android or iOS, nor a reason to install that documentation for those platforms.  Android requires position-independent code to even run, and will soon only allow shared libraries embedded in APKs.

So long story short, I didn't try to make --enable-android customizable, or test whether it is.

Reply:
Why is it called --enable-android if it is used for iOS as well?
We should use a more generic name, or create an alias for iOS.

Reply:
Sorry for the confusion there, this patch is purely about Android.  But I've been looking at iOS at the same time, and working to combine the Android and iOS setups as much as it makes sense.  I think in the end, we might end up with really similar setups with only a few quirks for the differences.

The Android and iOS builds currently are using a whole collection of custom flags.  That means that widely used build setups are unknown to the core Tor developers (unless you go digging for them).   My goal here is to get the main Android and iOS setups to build using a collective flag like --enable-android, so that the core Tor developers can add configurations there when they are ready to be used by the downstream projects (Tor Browser, Orbot, iCepa, OnionBrowser, etc).  For example, when code is split out into a module, like dirauth, and it is ready for mobile configurations, then that configure variable can be included in the Android and iOS build setups.

For example, iCepa is the mostly commonly used iOS Tor framework, here is how it calls ./configure: https://github.com/iCepa/Tor.framework/blob/30480101b3bd2166baced913347fc8bb7392b3da/Tor/tor.sh#L139

_tor-android-binary_ is the only Tor build that I know of that is currently included in shipping apps, here's how it calls configure:
https://github.com/guardianproject/tor-android/blob/0830c87a29ea197ecd9a0b5aee7d9a4fb88dce25/external/Makefile#L271

After this patch, it can be reduced to:
{{{
			./configure \
				--host=$(ALTHOST) \
				--enable-android \
				--enable-lzma \
				--enable-restart-debugging \
				--enable-static-libevent --with-libevent-dir=$(EXTERNAL_ROOT) \
				--enable-static-openssl --with-openssl-dir=$(EXTERNAL_ROOT) \
				--enable-zstd \
				--prefix=$(EXTERNAL_ROOT)
}}}

Reply:
> So long story short, I didn't try to make --enable-android customizable, or test whether it is.

I still can't agree with this.  This is not just about a change for android; this is about a change to the semantics of the configure.ac file, which will break the semantics of several other options unless we fix it.  It's never a great idea to break options silently, even if we're fairly sure we don't want to use them.

If you're sure that the options will never need change, then perhaps instead of --enable-android, it would make sense to have a "configure_for_android.sh" script in contrib/android/ or something?

Otherwise the fix is to move the "enable_android" check so that it happens before any of the other options that it adjusts.

Reply:
If you can point to something this breaks on non-Android platforms, I'll fix that.  Tor's ./configure has lots of issues when compiling for mobile, see some of the issues I've filed. And I've been told that the [pkg-config bug won't be fixed](https://trac.torproject.org/projects/tor/ticket/32191#comment:5), that affects the recommended Tor Android build config that is what all the apps are using.  Yet we are now talking about fixing hypothetical configurations.

The point of --enable-android is to provide a build that is as familiar to an Android developer as possible.  Android developers basically never use autotools, the majority are ignorant to the existence of autotools.  `ndk-build` and `cmake` are what's used, and these include automatic configurations like what's provided by --enable-android.   PIC/PIE is a [hard requirement](https://stackoverflow.com/questions/33659673/android-ndk-are-fpic-and-pie-mututally-exclusive), the code will not run without it. Android builds don't build HTML documentation, I've never seen an Android app build and include man pages, so failing on asciidoc will only confuse people.  "system torrc" is a UNIX concept, it does not belong in anyway on Android.   I'm also working on building Tor into Calyx's Android ROM, so I'm intimately familiar with what Android with tor provided by the system looks like.

See the configure lines I referenced that are currently used in the binaries shipped in Tor Browser, Orbot, Briar, Thali, Onion Browser, iCepa, etc.  And autotools have always been a nightmare for Android builds.  autotools is so centered on UNIX, and it really falls down when things are different.  (I've been working with autotools for 20 years).  So yes, this is weird when thinking in terms of UNIX and autotools, but it is an big improvement on the status quo.

Making Android builds is currently only possible with a brittle and elaborate wrapper.  I've gotten real Android builds working without the wrapper script in this patch for the first time.  I have been working to get Guardian Project out of the business of providing the official build system for Tor on Android.  Tor Browser uses our build system.  Integrations like this have to happen to achieve this.  And that will only happen if the tor maintainers are willing to accept that cross-compiling for mobile is really different than building on UNIX and Windows.

Reply:
**Other Tickets**

> And I've been told that the pkg-config bug won't be fixed, that affects the recommended Tor Android build config that is what all the apps are using.

I was asking questions on that ticket to try and work out how to fix the bug. Now that we know how to fix the issue, I have suggested a way forward on that ticket.

If there are issues with my suggestion, please reply on that ticket.

**This Pull Request**

> > Otherwise the fix is to move the "enable_android" check so that it happens before any of the other options that it adjusts.

@eighthave, what's wrong with this suggested fix?

As far as I understand, moving the check to the start will still set all the defaults that are required for Android. But if one of those defaults is wrong for a particular platform or build, then the developers will be able to change it.

Can you give a specific example where this solution will not work?

**Next Steps**

Can we please try to focus on the specific changes that we need to move forward on this pull request?

Tickets and pull requests aren't a great place for long-form discussion. If you want to talk about the broader context, and how tor can support mobile better, please write an email to the tor-dev list.

We're trying to help. But we do have engineering and compatibility constraints. If you want to change those constraints, let's talk about that more on the mailing list.

Reply:
I'm happy to focus on specifics. If it is enough to just move this block:
```diff
+dnl Enable minimal, standard Android build
+AC_ARG_ENABLE(android,
+     AS_HELP_STRING(--enable-android, [setup build for Android exeutable/shared library]))
+
+if test "x$enable_android" = "xyes"; then
+  asciidoc=false
+  enable_html_manual=no
+  enable_manpage=no
+  enable_pic=yes
+  enable_system_torrc=no
+  enable_tool_name_check=no
+  have_systemd=no
+fi
+
```

Right after `AC_CANONICAL_HOST`, then it could also enable a hard error when pkg-config is missing.  Otherwise, right before `AC_ARG_ENABLE(openbsd-malloc,` should work in theory.  It would need to be tested.

Reply:
Sorry if I'm grumpy in all this, all this fighting with cross-compiling, Android emulators, autotools vs NDK, etc. is driving me a little crazy.  But I can see the light at the end of the tunnel, though not always clearly...

Reply:
> it could also enable a hard error when pkg-config is missing

That's not how pkg-config works. It's an optional dependency that automatically sets some variables from *.pc files. But the user can always set those variables manually.

Reply:
pkg-config is currently not optional for the standard tor build on Android, e.g. with `--enable-lzma`, and in the case where pkg-config is missing, `./configure` will not exit with an error.  And for big wrapper scripts like are required for Android and iOS, that means the giant build process will complete but the resulting tor binary will not have lzma.  I fine with pkg-config being optional.  I still think it makes sense to skip pkg-config entirely for Android builds, and perhaps iOS, and just use --prefix to find dependencies.

Reply:
I can see that this lzma issue really concerns you. But I'm struggling to understand what you need us to change. Can you open another ticket with your request? I'm a bit confused by all the different topics in this one comment thread.

In any case, if you can get the right CFLAGS and LDFLAGS without pkg-config, then go for it! You can avoid installing pkg-config on android, or set all of its env vars to empty directories. And then you should be fine. No need to modify configure at all :-)

Reply:
so back to --enable-android, it needs to be something that stands out from the regular ./configure flags because how else can there be checks for requirements like `--host-` and cross-compiling?  The only other option I can see is basing all this stuff on:
```bash
   case "$host" in
       *-linux-android*) 
       # do checks here
       ;;
  esac
```